### PR TITLE
Ava 108/retry

### PR DESCRIPTION
**Pull request by kroos47**
Simple Retry for light client
PS: Exponential backoff is not yet implemented

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -129,155 +129,160 @@ pub async fn do_main() -> Result<()> {
...
-							log::info!(
-								"✅ Completed {} rounds of verification for block number {} ",
-								count,
-								num
-							);
-						} else {
-							log::info!("\n ❌ getting proof cells failed, data availability cannot be ensured");
-						}
+	//@TODO: better option than loop needed
+	// let ws_ = rpc::check_connection(cfg.full_node_ws.clone()).await?;
```
Comment (aterentic-ethernal) [inline src/main.rs:150]: I guess this is obsolete comment?

File: src/main.rs [hunk 2]
```diff
@@ -129,155 +129,160 @@ pub async fn do_main() -> Result<()> {
...
+	while let Some(z) = rpc::check_connection(&urls).await {
+		let (mut write, mut read) = z.split();
+		write
+			.send(Message::Text(
+				r#"{"id":1, "jsonrpc":"2.0", "method": "subscribe_newHead"}"#.to_string() + "\n",
+			))
+			.await
+			.context("ws-message(subscribe_newHead) send failed")?;
+
+		// let _subscription_result = read.next().await.unwrap().unwrap().into_data();
```
Comment (aterentic-ethernal) [inline src/main.rs:160]: I guess this is also obsolete comment?

File: src/main.rs [hunk 3]
```diff
@@ -129,155 +129,160 @@ pub async fn do_main() -> Result<()> {
...
+
+		// while(true) {
+		// 	match(read.next().await) {
+		// 		Ok(_) => {
+
+		// 		}
+		// 		Err(_) => {}
+		// 	};
+		// }
+
```
Comment (aterentic-ethernal) [inline src/main.rs:179]: I guess these also obsolete comments?

File: src/rpc.rs [hunk 4]
```diff
@@ -1,17 +1,21 @@
 extern crate futures;
 extern crate num_cpus;
 
+// use async_std::net::TcpStream;
+// use std::net::TcpStream;
```
Comment (aterentic-ethernal) [inline src/rpc.rs:5]: I guess this is also obsolete comment?

File: src/rpc.rs [hunk 5]
```diff
@@ -367,3 +371,54 @@ pub fn get_id_specific_size(num: Block) -> HashMap<u32, u32> {
...
+				break;
+			}
+		} else {
+			let client = hyper::Client::new();
+			let _res = match client.get(url_).await {
+				Ok(c) => c,
+				Err(_) => continue,
+			};
+			//@TODO: need to find an alternative way for http part
+			// if _res.status().is_success() {
```
Comment (aterentic-ethernal) [inline src/rpc.rs:416]: I guess this is also obsolete comment?

**Review by aterentic-ethernal — APPROVED** [review]

File: src/rpc.rs [hunk 6]
```diff
@@ -367,3 +371,54 @@ pub fn get_id_specific_size(num: Block) -> HashMap<u32, u32> {
...
+	// TODO: We are ignoring errors here, we should probably return result instead of option
+	for url in full_node_ws.iter() {
+		if let Ok((ws, _)) = connect_async(url).await {
+			return Some(ws);
+		};
+	}
+	None
+}
+
+pub async fn check_http(full_node_rpc: Vec<String>) -> Result<String> {
```
Comment (sh3ll3x3c) [inline src/rpc.rs:394]: We should make an effort to either create more descriptive function/variable names or add explanatory comments (ideally, both).

Comment (kroos47) [reply src/rpc.rs:394]: Added some comments for the functions

File: src/rpc.rs [hunk 7]
```diff
@@ -367,3 +369,55 @@ pub fn get_id_specific_size(num: Block) -> HashMap<u32, u32> {
...
+//parsing the urls given in the vector of urls
+pub fn parse_urls(urls: Vec<String>) -> Result<Vec<url::Url>> {
+	urls.iter()
+		.map(|url| url::Url::parse(url))
+		.map(|r| r.map_err(|parse_error| anyhow!("Cannot parse URL: {}", parse_error)))
+		.collect::<Result<Vec<_>>>()
+}
+
+//fn to check the ws url is working properly and return it
+pub async fn check_connection(
```
Comment (sh3ll3x3c) [inline src/rpc.rs:381]: Are we checking the entire provided list of wss if they are connectable?

Comment (kroos47) [reply src/rpc.rs:381]: yes we are! if one connection is successfull then we return that and if the same is lost then we again try to connect to other one!

File: src/rpc.rs [hunk 8]
```diff
@@ -367,3 +369,55 @@ pub fn get_id_specific_size(num: Block) -> HashMap<u32, u32> {
...
+	}
+	None
+}
+
+//fn to check the rpc_url is secure or not and if it is working properly to return
+pub async fn check_http(full_node_rpc: Vec<String>) -> Result<String> {
+	let mut rpc_url = String::new();
+	for x in full_node_rpc.iter() {
+		let url_ = x.parse::<hyper::Uri>().context("http url parse failed")?;
+		if is_secure(x) {
```
[the remark this answers is not in the document]
Comment (kroos47) [reply src/rpc.rs:398]: So then we can remove and modify this part, but right now this check is needed for verifying. We can remove when the HTTPS redirect is there?

[the remark this answers is not in the document]
Comment (sh3ll3x3c) [reply src/rpc.rs:398]: if it works, let's leave it for now, but let's keep in mind this issue in the future and how it's handled on the node

**Review by sh3ll3x3c — APPROVED** [review]

**Comment by sh3ll3x3c** [comment]
Let's just squash merge, too many commits @kroos47 .

**Pull request merged**